### PR TITLE
fix: clipboard get the different unicode whitespace

### LINKED ISSUE #4225: [BUG] - `@nextui-org/snippet` Component Adds Extra Characters When Copying Code Snippets
### NextUI Version

2.2.9

### Describe the bug

The `Snippet` component from `@nextui-org/snippet` adds invisible characters (e.g., non-breaking spaces) when using the built-in copy functionality. This results in formatting issues and errors when the copied code is pasted into a code editor. For instance, JSON code becomes invalid with errors like `"Property keys must be doublequoted"` due to extra hidden characters.

### **Actual Behavior**
The copied code includes extra invisible characters, likely non-breaking spaces (`\u00A0`), which result in:
- Errors in JSON files such as `"Property keys must be doublequoted"`.
- Code editors displaying red squiggly lines or warnings.

For example, the pasted code looks like this (with invisible characters causing issues):
```json
{
  "name": "Next.js PWA",
  "short_name": "NextPWA",
  "description": "A Progressive Web App built with Next.js and React",
  "start_url": "/",
  "display": "standalone",
  "background_color": "#ffffff",
  "theme_color": "#000000",
  "icons": [
    {
      "src": "/icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "/icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ]
}
```

Notice how some spaces are incorrectly formatted (non-breaking spaces).

### **Technical Analysis**
The issue appears to be related to how the `Snippet` component processes and renders the code internally. During the copy operation:
1. Non-breaking spaces (`\u00A0`) or other hidden formatting characters are added.
2. These characters are pasted into editors, causing JSON parsing and validation errors.

---

### **Proposed Solution**
1. **Sanitize the Copied Content**:
   - Strip non-breaking spaces or hidden formatting characters before copying to the clipboard.
   - Use `.replace(/\u00A0/g, " ")` or equivalent sanitization.

2. **Allow Customization**:
   - Provide a prop (e.g., `onCopy`) that lets users customize the copy functionality or sanitize the copied content.

3. **Update Documentation**:
   - Warn users about potential formatting issues in JSON or code snippets when using the `Snippet` component.

---

### **Environment**
- **Next.js**: Version `14.4.0`
- **Browser**: Chrome Version `119.0.0`
- **Operating System**: Windows 11

### Your Example Website or App

_No response_

### Steps to Reproduce the Bug or Issue

1. **Set up a Next.js project**:
   ```bash
   npx create-next-app@latest my-next-app
   cd my-next-app
   npm install @nextui-org/react
   ```
2. **Use the `Snippet` component** as shown in the example below:
   ```jsx
   import { Snippet } from "@nextui-org/snippet";

   const codeSnippet = `
   {
     "name": "Next.js PWA",
     "short_name": "NextPWA",
     "description": "A Progressive Web App built with Next.js and React",
     "start_url": "/",
     "display": "standalone",
     "background_color": "#ffffff",
     "theme_color": "#000000",
     "icons": [
       {
         "src": "/icon-192x192.png",
         "sizes": "192x192",
         "type": "image/png"
       },
       {
         "src": "/icon-512x512.png",
         "sizes": "512x512",
         "type": "image/png"
       }
     ]
   }
   `;

   export default function App() {
     return <Snippet showCopyButton>{codeSnippet}</Snippet>;
   }
   ```
3. **Run the app locally**:
   ```bash
   npm run dev
   ```
4. **Copy the code** using the copy button in the `Snippet` component.
5. **Paste the copied code** into any code editor (e.g., VS Code).

### Expected behavior

The copied code should match the original snippet exactly, with no additional characters or formatting artifacts.

### Screenshots or Videos

image

### Operating System Version

- OS: Windows 11

### Browser

Chrome

### PR DESCRIPTION
**Pull request by winchesHe**
Closes #4225

## Summary by CodeRabbit

- **New Features**
	- Introduced a utility function to normalize whitespace in clipboard operations.
	- Updated the snippet component to provide installation commands for the `@nextui-org/react` package.
- **Bug Fixes**
	- Resolved issues with clipboard handling of different Unicode whitespace characters.

### CODE REVIEW

File: .changeset/rich-moles-compare.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@nextui-org/use-clipboard": patch
+"@nextui-org/snippet": patch
```
Comment (wingkwong) [inline .changeset/rich-moles-compare.md:3]: not required (?)

Comment (winchesHe) [reply .changeset/rich-moles-compare.md:3]: if the use-clipboard upgrade to 2.1.2 then the latest @nextui-org/snippet is using 2.1.1 will it upgrades too? or we can delete it

Comment (wingkwong) [reply .changeset/rich-moles-compare.md:3]: `@nextui-org/snippet` has a dependency on `use-clipboard`. When `use-clipboard` got bumped, `@nextui-org/snippet` will be also bumped automatically by changeset.

File: packages/hooks/use-clipboard/src/index.ts [hunk 2]
```diff
@@ -33,18 +33,26 @@ export function useClipboard({timeout = 2000}: UseClipboardProps = {}) {
     [onClearTimeout, timeout],
   );
 
+  const transformWhitespace = useCallback((text: string) => {
+    // Manually replace all whitespace to avoid get different unicode characters;
+    return text.replace(/\s/g, " ");
```
Comment (wingkwong) [inline packages/hooks/use-clipboard/src/index.ts:38]: Let's say if my input is a json format (says it got 20 lines), I copy it and paste to my editor. The pasted value will be just one line, which is not expected. Current behaviour is the paste value is same as how it looks as input.

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/hooks/use-clipboard/src/index.ts (hunk 2 above)
Comment (winchesHe) [reply packages/hooks/use-clipboard/src/index.ts:38]: can you send me the issue input, i will handle it

Comment (wingkwong) [reply packages/hooks/use-clipboard/src/index.ts:38]: Retrieved from the reported issue

```json
{
  "name": "Next.js PWA",
  "short_name": "NextPWA",
  "description": "A Progressive Web App built with Next.js and React",
  "start_url": "/",
  "display": "standalone",
  "background_color": "#ffffff",
  "theme_color": "#000000",
  "icons": [
    {
      "src": "/icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "/icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ]
}
```

**Comment by jrgarciadev** [comment]
@winchesHe @wingkwong  is this PR ready to merge?

**Comment by wingkwong** [comment]
@jrgarciadev there is a case that need to handle.

**Comment by winchesHe** [comment]
> @jrgarciadev I think not. There is a case that need to handle.

Updated

**Review by wingkwong — APPROVED** [review]

**Pull request merged**